tests/validator/validator.cpp

```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(2), validator->issueCount());
+    EXPECT_EQ_ISSUES(errorMessagesValidator, validator);
+}
+
+TEST(Validator, importComponentWithInvalidName)
+{
+    const std::vector<std::string> errorMessages = {
+        "Import of component 'c' has an invalid URI in the xlink:href attribute.",
+        "The attempt to resolve imports with the model at '"
+            + resourcePath("importer/")
+            + "i am broken and invalid.cellml' failed: the file could not be opened.",
+    };
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/invalid_import_url.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessages[0], validator->issue(0)->description());
+    EXPECT_EQ(model->component("c")->importSource(), validator->error(0)->importSource());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(1), importer->errorCount());
+    EXPECT_EQ(errorMessages[1], importer->issue(0)->description());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessages[0], validator->issue(0)->description());
+    EXPECT_EQ(model->component("c")->importSource(), validator->error(0)->importSource());
+}
+
+TEST(Validator, importSecondGenComponentWithInvalidUri)
+{
+    const std::string errorMessageValidator =
+        "Imported component 'c' is not valid because:\n"
+        "  -> Component 'c' importing 'c' from 'invalid_import_url.cellml' has an error:\n"
+        "   - Import of component 'c' has an invalid URI in the xlink:href attribute.";
+    const std::string errorMessageImporter =
+        "The attempt to resolve imports with the model at '"
+        + resourcePath("importer/")
+        + "i am broken and invalid.cellml' failed: the file could not be opened.";
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/import_invalid_import_url.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->issueCount());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(errorMessageImporter, importer->issue(0)->description());
+    EXPECT_EQ(model->component("c", true), importer->issue(0)->component());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessageValidator, validator->issue(0)->description());
+}
+
+TEST(Validator, highIndexUnitsImport)
```

Comment:
The name of the test doesn't fit the test. What about `importedUnitBasedOnNonExistingUnit` instead?

Reply:
1f87e46